python/ray/serve/tests/test_standalone.py

```diff
@@ -128,5 +129,48 @@ def test_middleware():
     ray.shutdown()
 
 
+@pytest.mark.skipif(
+    not hasattr(socket, "SO_REUSEPORT"),
+    reason=("Port sharing only works on newer verion of Linux. "
+            "This test can only be ran when port sharing is supported."))
+def test_cluster_handle_affinity():
+    cluster = Cluster()
+    # HACK: using two different ip address so the placement constraint for
+    # resource check later will work.
+    head_node = cluster.add_node(node_ip_address="127.0.0.1", num_cpus=4)
+    cluster.add_node(node_ip_address="0.0.0.0", num_cpus=4)
+
+    ray.init(head_node.address)
+
+    # Make sure we have two nodes.
+    node_ids = [n["NodeID"] for n in ray.nodes()]
+    assert len(node_ids) == 2
+
+    # Start the backend.
+    client = serve.start(http_port=randint(10000, 30000), detached=True)
+    client.create_backend("hi:v0", lambda _: "hi")
+    client.create_endpoint("hi", backend="hi:v0")
+
+    # Try to retrieve the handle from both head and worker node, check the
+    # router's node id.
+    @ray.remote
+    def check_handle_router_id():
+        client = serve.connect()
+        handle = client.get_handle("hi")
+        return get_node_id_for_actor(handle.router_handle)
+
+    router_node_ids = ray.get([
+        check_handle_router_id.options(resources={
+            node_id: 0.01
+        }).remote() for node_id in ray.state.node_ids()
+    ])
+
+    assert set(router_node_ids) == set(node_ids)
```

Comment:
Because we're converting to `set`, it seems like this assertion could pass even if the routers somehow got swapped.  Is this something to worry about?